README.md
# Kiln

Kiln is a tool for both baking and monitoring on the Tezos network. It
provides a locally hosted graphical interface, binaries for
tezos-client, tezos-node, tezos-baker, and tezos-endorser, and it
builds a cache of chain data from the nodes to which it connects.

**For step-by-step instructions on how to get started, see the
Obsidian System's medium post [How to Install Kiln and Bake on
Ubuntu](https://medium.com/@obsidian.systems/how-to-install-kiln-and-bake-on-ubuntu-a13d17df63c).**

**Installation guide for MacOS Catalina or later is available [here](https://medium.com/tezos-kiln/how-to-install-kiln-on-macos-catalina-ce0821f97dcf).**

Past release notes are available
[here](https://medium.com/@obsidian.systems).

If you don't need/want graphical user interface, consider baking
setup with [Serokell Tezos
distributions](https://github.com/serokell/tezos-packaging/blob/master/docs/baking.md).

## Quick Start

- Download latest release from [Releases](https://gitlab.com/tezos-kiln/kiln/-/releases)

### Ubuntu

- Install or update:

    **On desktop:**
    - Double click downloaded .deb file
    - Click `Install` button

    **From the command line:** assuming .deb file
    (e.g. `kiln_0.9.3_amd64.deb`) is downloaded to `~/Downloads` folder
    - Open terminal
    - Run

```
sudo dpkg -i ~/Downloads/kiln_0.9.3_amd64.deb
```

- Go to Kiln web user interface: open <http://localhost:8000> in a web browser

### MacOS

- Install or update:
  - Right click downloaded .pkg file, select "Open With > Installer"
  - Confirm you would like to install this application
  - Follow installer prompts

- After computer reboots, go to Kiln web user interface: open
  <http://localhost:8444> in a web browser

## Troubleshooting

### UI stops updating

There are instances where Kiln’s user interface will stop updating,
but the backend will continue functioning properly. For instance, resolving a
notification will not cause the notification to disappear until the
page is refreshed. Stopping and restarting Kiln fixes this issue.

Restart Kiln (Ubuntu):

```
sudo systemctl stop kiln
sudo systemctl start kiln
```

Restart Kiln (MacOS):

```
launchctl stop tezos.kiln
launchctl start tezos.kiln
```

### Node status in UI is stale

Sometimes Kiln UI may [stop properly updating monitored node
status](https://gitlab.com/tezos-kiln/kiln/-/issues/4).

To verify that Kiln's Tezos node is synchronized with the network run
the following command:

Check node is synced (Ubuntu):

```
/usr/share/kiln/nix/store/*-tezos-*/bin/tezos-client --endpoint http://localhost:8733 bootstrapped
```

Check node is synced (MacOS):

```
export DYLD_FALLBACK_LIBRARY_PATH=/usr/local/kiln-nix/lib
/usr/local/kiln-nix/nix/store/*tezos*/bin/tezos-client --endpoint http://localhost:8733 bootstrapped
```

If node is synchronized with the network, the command will print "Node
is bootstrapped." message and exit immediately.

Or check via node's web API, e.g. using curl:

```
curl http://localhost:8733/chains/main/is_bootstrapped
```

Healthy node should return
```
{"bootstrapped":true,"sync_state":"synced"}
```

Stopping and restarting Kiln typically resolves the issue.

### Missed bake/endorsement
A typical reason for missing a bake or endorsement is unavailable
Ledger device. Examine baker and endorser logs to verify if that's the
case.

Check logs (Ubuntu):

```
# everythin
journalctl -u kiln

# filter messages from baker
journalctl -u kiln | grep baker

# filter message from endorser within a time period
journalctl -u kiln --since "5 days ago" --until "1 hour ago" | grep endorser
```

Consult journalctl
[documentation](https://manpages.ubuntu.com/manpages/cosmic/man1/journalctl.1.html)
or
[tutorial](https://www.digitalocean.com/community/tutorials/how-to-use-journalctl-to-view-and-manipulate-systemd-logs)
for more ways to browse and filter the logs.

Check logs (MacOS):

- go to `Applications > Utilities` and launch `Console` application.
- Select `Log Reports`
- Type `kiln` in search bar
- Select one of kiln log files

Verify that Ledger is connected (Ubuntu):

```
/usr/share/kiln/nix/store/*-tezos-*/bin/tezos-client --endpoint http://localhost:8733 list connected ledgers
```

Verify that Ledger is connected (MacOS):

```
export DYLD_FALLBACK_LIBRARY_PATH=/usr/local/kiln-nix/lib
/usr/local/kiln-nix/nix/store/*tezos*/bin/tezos-client --endpoint http://localhost:8733 list connected ledgers
```

If ledger is not listed in the output:

- Make sure Tezos baking application is running on the device
- Make sure that device is properly plugged in, on both ends
- Try unplug the device, plug it back in, unlock and start baking app again
- Try different USB port
- Try different USB cable

### Adding Ledger indicator to the UI

To add a ledger indicator to your UI to check connectivity every 60
seconds do the following:

On Ubuntu:

```
sudo su -
cd /var/lib/kiln/exe-dir/
mkdir -p config
echo 60 > config/ledger-check-delay
systemctl restart kiln
```

On MacOS

```
cd ~/Library/Kiln
mkdir -p config
echo 60 > config/ledger-check-delay
launchctl stop tezos.kiln
launchctl start tezos.kiln
```

Indicator should now appear in the top bar of your UI.

## System Requirements

System requirements are dependent on whether you plan on running a
Tezos node in Kiln, which is necessary for baking.

*Note: If you are running a node using our Docker Image, note that
Docker's default memory allocation is 2GB. Increasing this to 8GB is
highly recommended.*

### System Requirements For Monitoring

**Disk:** The monitor uses PostgreSQL for *all* storage. The entire
database typically uses about 1-2GB.

**Memory:** Idle memory usage is typically under 1GB. When
initializing history (usually right after start-up or adding your
first node) memory can spike to about 3GB for a short time.

**CPU:** Running with at least 2 cores is recommended.

### System Requirements For Baking

Kiln bakes with a local node, which increases system requirements.

**Disk:** A node running in Kiln will sync with the Tezos blockchain,
which is currently ~70GB. **SSD is highly recommended over HHD.**

**Memory:** Recommendated RAM for running a Tezos Node is 8GB. We
recommend at least 10GB RAM to account for the node, baker, endorser,
and Kiln’s processes.

**CPU**: Running with at least 2 cores is recommended.

## Using Kiln to Bake

Kiln, in conjunction with [Tezos Baking for the Ledger Nano
S](https://github.com/obsidiansystems/ledger-app-tezos), can be used
to bake. Kiln runs the node, baker, and endorser locally while
monitoring them to notify the user of common issues and events.

Baking requires the Kiln node to be fully synced with the Tezos
blockchain. Once the node is synced, click ‘Add Baker’ in the left
sidebar. Then click ‘Start Baking’ and follow the instructions
provided.

Once you’ve registered as a delegate, it will take time for you to
earn baking and endorsing rights. As soon as Kiln detects you have
rights, your next right will be displayed.

## Using Kiln as a Monitor

Kiln monitors nodes, bakers, and the Tezos network to keep bakers
fully informed. The dashboard displays a tile with relevant
information for all the nodes and bakers it is monitoring, and it
alerts the user of any issues that it may encounter.

### Network Monitoring

Kiln produces a notification if an update is pushed to the Tezos
network. For instance, if a user is monitoring mainnet with Kiln, they
will be notified if mainnet is updated.

### Node Monitoring

Kiln produces a notification if a monitored node:

- Is on the wrong network
- Is not on the fittest branch
- Falls behind the current head block level
- Cannot be reached by the Monitoring Software (e.g. is offline)
- Reports fewer than a specified number of active peer connections

### Baker Monitoring

Kiln produces a notification if a monitored baker:

- Misses a baking or endorsing opportunity
- Is accused of double baking or double endorsing
- Has been deactivated due to inactivity or will be within one cycle

*Note: To monitor a baker, you must be monitoring a node. Public Nodes
provide general information about the network, but they do not provide
full block history or baking and endorsing rights, which are required
for monitoring a baker.*

### Notification Pathways

In addition to in-app alerts, users can configure Kiln to send
Telegram alerts or use their SMTP Mail Server to send alerts to the
email addresses of their choice. Notification pathways can be
configured on the Settings page.

For more information, see notes on configuring
[Telegram](#configuring-telegram-notifications) and
[Email](#configuring-email-notifications) notifications below.

## Initial Setup

### Running a node

Click *Add Node* in the left panel. Then click *Start Node* on the
left side of the modal. Kiln will generate an identity for a local
node with the RPC port 8733 and immediately begin syncing with the
blockchain. If this port is already in use, Kiln will have difficulty
monitoring this node (the port can be modified by [command line
options](./docs/distros/build-from-source.md#command-line-options)). The Kiln Node can be stopped or
restarted through the options menu on the Node's tile on the
Dashboard.

### Adding monitored nodes

Click *Add Node* from the left panel and enter the address of the node
you would like to monitor under *Monitor via Address*. Then click *Add
Node*. For example, if you would like to add a local node with an RPC
interface on the default port of `8732`, you can enter
`http://localhost:8732`\*. You can add any node URL to which you know
the RPC address. If you do not know the RPC address of the node or if
the node wasn't started with `--rpc-addr`, the monitor will not be
able to retrieve information from the node.

\* If you're running the monitor from Docker on macOS, `localhost`
will not point to your *host*'s network. Instead you can use
`host.docker.internal` instead of `localhost`,
e.g. `http://host.docker.internal:8732`. `localhost` will also not
work on Linux if you run the container without `--network host`, but
you can't use `host.docker.internal` in this case.

Once you’ve added at least one node or a Public Node, the Dashboard
header will show the network status.

*Note: This assumes that you are running at least one Tezos
node. See [Tezos
documentation](https://tezos.gitlab.io/introduction/howtoget.html) on
various ways of running a node*

### Adding public nodes

Click *Add Node* from the left panel and click one of the tiles under
*Connect to a Public Node*. Clicking again will disable the Public
Node.

### Monitoring a baker

Click *Add Baker* from the left panel and input the public key hash
(PKH) of the baker you would like to monitor. Kiln will then use
Monitored Nodes to gather information about that baker from the
blockchain. This initial query can take up to a few hours.

*Note: To monitor a baker, you must be monitoring a node - public
nodes are not sufficient. Public Nodes provide information about the
current head block, but they do not provide full block history of
baking and endorsing rights.*

### Configuring Telegram notifications

Click *Settings* from the left panel then click *Connect Telegram* and
follow the instructions in the popup. Telegram notifications can be
sent to a group or directly to a single user.

### Configuring email notifications

Click *Settings* from the left panel and provide the SMTP
configuration for your SMTP server in the form under *Email*. Add an
email address to receive alerts and click *Save Settings*.

## Contact Us

Users can join Tezos Baking Slack (by emailing for an invite at
joinslack@tezos-kiln.org) to provide feedback.


CHANGELOG.md
# Changelog

## 0.8.1

  * Bump Tezos mainnet node to include fixes to the mempool and the Tezos node
    (f0607f4b1f8c0f2a0a48adb4bc76378406203714)

### Known Issues

  * Kiln's memory usage may spike at the startup for few minutes.
    It is recommended to have swap memory enabled to avoid causing issues, especially if the tezos-node is also running on the same machine.
    Disabling the OS Public Node via the config file could reduce the Kiln's memory consumption further.

## 0.8.0

  * Updated tezos node dependencies to support the Carthage protocol and the new, more compact storage backend (irmin-pack).

  * Updated the Kiln Node management code to require the new tezos node, with a prompt to delete the old node and to initialize a new one if the existing data store is not at least version 0.0.4.

  * Tweaked the protocol constant parsing to support Carthage.

  * Support for the Carthagenet test network. (Babylonnet is still available for the time being.)

## 0.7.4

  * Ledger connectivity check is now disabled by default.
    To enable the check use the 'ledger-check-delay' option.

  * Bug Fixes
    * Fix the export logs functionality for Debian.

## 0.7.3

  * Ledger connectivity icon and alert
    Kiln now displays an icon on the top bar to indicate if the Ledger is connected/disconnected,
    and also gives an alert to the user if the Ledger device gets disconnected.

  * Support for baking with the new ed25519 derivation scheme (bip25519)

  * Support for babylonnet. (alphanet has been removed)

  * Snapshot import can be cancelled.

  * Improved log messages.
    Kiln now gives better log messages when it is not able to obtain required information from available nodes.

  * Support building Kiln on MacOS from source code using nix

  * "New Tezos software version is available" notification logic has been improved.
    Kiln now detects the version of each monitored node separately, and notify on the node tile UI if the node is running the software different from the latest version.
    This feature can also be disabled / configured via command line options.

  * TzScan public node has been removed.

  * Block hashes on node tile now links to tzstats.com (previously tzscan.com was used).

  * Bug Fixes
    * In some cases Kiln would continue to show "Gathering baking data", even though it has obtained all the available data (of preserved cycles) from the nodes.
      This has been fixed.
    * Fixed baker alert logic to not show the alert if the network is changed.
    * Babylon specific
      * Block's operation parsing now works if the entrypoint is an empty string.
      * Double endorsement detection logic has been fixed.

  * UI improvements
    * When adding a new baker, the Public Key Hash's curve and derivation info is shown as a hover tool tip.
    * Hovering over 'Network' in Kiln's header shows the current protocol.

  * Code improvements
    * The underlying app frameworks used by Kiln (Rhyolite, Obelisk and Reflex) have been upgraded.
    * The core logic of Tezos block data and node RPC communications, written in Haskell, has been moved to a separate project (https://gitlab.com/obsidian.systems/tezos-bake-monitor-lib)
      This will allow anyone to easily make use of this code to create other projects for Tezos.

### Known Issues
  * The copy button on the 'Network' -> 'Current Protocol' pop up might fail to copy the protocol hash to the clipboard.
  * On cancelling snapshot import the UI might display "Snapshot import failed" error.

## 0.7.2

  * Include the tezos-node containing: Mempool: remove parsing exceptions (a7d357bb)

    https://gitlab.com/tezos/tezos/commit/a7d357bb290eab43c92ec2245811e9af959d1d4c

### Known Issues
    * Upgrading Kiln from v0.6.* on Ubuntu/Debian overwrites the /etc/kiln/args file
    * Kiln will not work on zeronet.

## 0.7.1

  * Support for Babylon protocol (005_PsBabyM1)
    * This release has a fix to force-run the PsBabyM1 baker/endorser binaries.
      This is required to avoid missing the bake/endorse opportunity for the first block of the new protocol.

    * With 0.7.0 users can miss bake/endorse opportunities for first couple of blocks in the new protocol.
      It is recommended to upgrade to 0.7.1 before the start of Babylon protocol.

  * This removes the BABY5H baker/endorser, so kiln will not work on zeronet.

### Known Issues
    * Upgrading Kiln on Ubuntu/Debian overwrites the /etc/kiln/args file

## 0.7.0

  * Support for Babylon protocol (005_PsBabyM1 and 005_PsBABY5H)
    * Kiln now has tezos binaries to do baking operations on Athens (004) and the new 005 protocols.
    * Fixes for changes in the format of operations. This is required to properly parse block data in Kiln.
    * Fixes for changes in RPC. This enables Kiln to properly communicate with tezos-node running on new protocol.

  * Support for varying protocols.
    * Kiln now understands that various parameters of the blockchain can change over time due to amendement process.
      This is essential to gather data from blocks of past and current protocols.

  * Bug Fixes
    * Obsidian Public Node APIs
    * UI - Baker alerts now resolve on the tile without a page refresh.

  * Supported networks
    * alphanet, mainnet, zeronet (babylonnet is not supported)

### Known Issues
    * Upgrading Kiln on Ubuntu/Debian overwrites the /etc/kiln/args file

## 0.6.2

  * Update mainnet tezos binaries. This fixes the problem of 100% cpu usage by the tezos-node (and "updater: COMPILATION ERROR" messages). The tezos-node released with v0.6.1 has problem in compiling the new protocol binaries (PsBABY5HQTSk), and this issue surfaced now because mainnet entered the testing phase of the protocol amendement cycle.

## 0.6.1

  * Export logs for Kiln node, baker, and endorser via UI on Ubuntu/Debian.
  * Fix 'Add Baker' flow issues when communicating with tezos-client.
  * Performance improvements and bug fixes for alerts.

tezos-bake-monitor/src/BakeMonitor.hs
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}
{-# OPTIONS_GHC -fno-warn-unused-do-bind #-} -- for the parsers specifically

import Control.Concurrent
import Control.Concurrent.Async
import Control.Lens.Combinators (views, over, set, _head)
import Control.Monad
import Control.Monad.Reader
import Data.Aeson (encode)
import Data.Attoparsec.Text hiding (take, option)
import qualified Data.Attoparsec.Text as P
import Data.Char
import Data.Monoid ((<>), mempty)
import Data.Text (Text)
import Data.Time
import Data.Word
import GHC.IO.Exception
import Network.HTTP.Client
import Network.HTTP.Client.TLS
import Options.Applicative hiding (Parser)
import qualified Options.Applicative as O
import Snap hiding (method)
import System.Process
import Text.Read
import qualified Data.Text as T
import qualified Data.Text.IO as T
import qualified Data.Text.Lazy.IO as LT

import Tezos.BakeMonitor.Types
import Tezos.NodeRPC

-- TODO: write a pid file for the spawned baker client

-- Type for common sorts of categories of message that the baker emits while baking.
-- We're mainly interested in counting the blocks that are injected, but some of the rest is potentially useful.
data MessageType =
    MessageType_Selected BlockPrefix  -- Select candidate block after BKidogWLoxoM (slot 1) fitness: 00::00000000000000df
  | MessageType_Injected BlockPrefix Word64 {- level -} -- Injected block BKiNQABfPLcg for my-ident after BKiNpAqXuqEx  (level 222, slot 0, fitness 00::00000000000000de, operations 0+0+0+0)
  | MessageType_NoNonce BlockPrefix   -- No nonce to reveal for block BKiNmhYodVSR
  | MessageType_Error      -- Error while endorsing:
  | MessageType_ErrorCont  -- Error, dumping error stack:
                           --   Wrong predecessor BKiQtCxSQRGcr3QPX7RR62VNnM9DXNv4NnJ6HjJso3f9W87FLHT, expected BKiUhCVyeftvggKDw3UjHhXSXdmUnV4epidsyHnS5B2XEA2gcEh
                           -- ^^ note these two spaces.
  | MessageType_Unknown    -- Anything else.
  deriving Show

blockPrefix :: Parser BlockPrefix
blockPrefix = BlockPrefix <$> P.takeWhile isAlphaNum

errorCont :: Parser MessageType
errorCont = do
  string "Error, dumping error stack:" <|> string "  "
  return MessageType_ErrorCont

errorLine :: Parser MessageType
errorLine = do
  string "Error"
  return MessageType_Error

selectCandidate :: Parser MessageType
selectCandidate = do
  string "Select candidate block after "
  bp <- blockPrefix
  return $ MessageType_Selected bp

injectedBlock :: Parser MessageType
injectedBlock = do
  string "Injected block "
  bp <- blockPrefix
  P.takeWhile (/= '(')
  string "(level "
  n <- decimal
  return $ MessageType_Injected bp n

noNonce :: Parser MessageType
noNonce = do
  string "No nonce to reveal for block "
  bp <- blockPrefix
  return $ MessageType_NoNonce bp

messageType :: Parser MessageType
messageType = errorCont <|> errorLine <|> selectCandidate <|> injectedBlock <|> noNonce

classify :: Text -> MessageType
classify t = case parseOnly messageType t of
  Left _ -> MessageType_Unknown
  Right x -> x

fetchBlockFromFragment :: Text -> Manager -> BlockPrefix -> IO BlockHash
fetchBlockFromFragment nodeAddr httpMgr = runNodeRPCT (NodeRPCContext httpMgr nodeAddr) . go
  where
    go pfx = nodeRPC (Complete pfx) >>= \case
        RpcResponse_HttpException e -> error $ "bad response from node" <> show e <> "for prefix" <> show pfx
        RpcResponse_UnexpectedStatus s -> error $ "bad response from node" <> show s <> "for prefix" <> show pfx
        RpcResponse_NonJSON clue raw -> error $ "Non JSON response from node: " <> show clue <> "\n" <> show raw <> "for prefix" <> show pfx
        RpcResponse_Success v -> case v of
            (blockId:_) -> return blockId
            _ -> error $ "Block Prefix not known to node" <> "for prefix" <> show pfx

opts :: O.Parser (IO ())
opts = mainArgs
  <$> option auto
      (  long "port"
      <> short 'p'
      <> help "Port to listen on"
      <> showDefault
      <> value 9800
      <> metavar "PORT"
      )
  <*> strOption
      (  long "rpchost"
      <> short 'r'
      <> help "Node url to listen on (without protocol)"
      <> showDefault
      <> value "127.0.0.1:8732"
      <> metavar "NODERPC"
      )
  <*> strOption
      (  long "client"
      <> short 'c'
      <> help "Which tezos-client executable to use"
      <> metavar "FILE"
      )
  <*> strOption
      (  long "identity"
      <> short 'i'
      <> help "Which identity to use for baking"
      <> metavar "TEZOS-IDENTITY"
      )

main :: IO ()
main = join $ do
  customExecParser
    (prefs $ showHelpOnEmpty <> showHelpOnError)
    (info (opts <**> helper) idm)

bumpBlockSeen :: UTCTime -> Report -> Report
bumpBlockSeen now = over report_lastSeen (\old -> max (Just now) old)

mainArgs :: Int -> Text -> FilePath -> String -> IO ()
mainArgs monitorPort nodeRPCLocation clientExecutable identity = do
  let [rpcAddr, rpcPort] = T.splitOn ":" nodeRPCLocation
  (_, Just out, Just err, ph) <- createProcess
    (proc clientExecutable ["--addr", T.unpack rpcAddr, "--port", T.unpack rpcPort, "launch", "daemon", identity, "-B", "-E", "-D"])
      { std_out = CreatePipe
      , std_err = CreatePipe
      }
  dataRef <- newMVar $ Report
    { _report_counts = Count 0 0 0
    , _report_lastBaked = []
    , _report_errors = []
    , _report_failedBaker = []
    , _report_lastSeen = Nothing
    , _report_tezzies = Nothing
    }

  let updateData f = modifyMVar_ dataRef $ return . f

  httpMgr <- liftIO $ newManager tlsManagerSettings

  -- TODO: if the baker is on a bad fork, it may erroneously belive it has
  -- more tezzies than it really does on the real fork.  this should be moved
  -- to the central monitor, and checked by several nodes
  void . forkIO $ forever $ do
    balanceLine <- readProcess clientExecutable ["get", "balance", "for", identity] ""
    let balance = readMaybe (filter (\c -> isDigit c || c == '.') balanceLine)
    updateData (set report_tezzies $ fmap Tezzies balance)
    threadDelay (60*10^(6 :: Int))

  -- consume from stdout looking for data.
  void . forkIO $ forever $ do
    msg <- T.hGetLine out
    now <- getCurrentTime
    case classify msg of
      MessageType_Selected _ -> modifyMVar_ dataRef $ \tops -> do
        return . bumpBlockSeen now . over (report_counts . count_selected) (+1) $ tops
      MessageType_Injected h level -> do
        modifyMVar_ dataRef $ \tops -> do
          blockHash <- fetchBlockFromFragment ("http://" <> nodeRPCLocation) httpMgr h
          let bakedV = Baked
                { _baked_seq = views (report_counts . count_injected) (+1) tops
                , _baked_hash = blockHash
                , _baked_time = now
                , _baked_level = level
                }
          return
            . over (report_counts . count_injected) (+1)
            . over report_lastBaked (\bs -> take 20 $ bakedV : bs)
            . bumpBlockSeen now
            $ tops
      MessageType_Error ->
        updateData
          $ over (report_counts . count_errors) (+1)
          . over report_errors (take 20 . (Error now msg :))
      MessageType_ErrorCont ->
        updateData $ over (report_errors . _head . error_text) (`T.append` msg)
      MessageType_NoNonce _ -> do
        updateData (bumpBlockSeen now)
      MessageType_Unknown -> return ()

    T.putStrLn (T.pack (show now) <> ": " <> msg)
    print (classify msg)

  let bakerTask = waitForProcess ph >>= \case
        ExitSuccess -> error "this is really not supposed to happen..."
        _ -> modifyMVar_ dataRef $ \tops -> do
          stdErrors <- LT.hGetContents err
          LT.putStr stdErrors
          return $ over report_failedBaker (stdErrors:) tops

  let rootHandler = do
        c <- liftIO $ readMVar dataRef
        writeLBS (encode c)

  let httpTask = httpServe (setPort monitorPort mempty) rootHandler
  void $ race bakerTask httpTask
